sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala

```diff
@@ -1870,4 +1870,19 @@ class SQLQuerySuite extends QueryTest with SharedSQLContext {
       assert(sampled.count() == sampledOdd.count() + sampledEven.count())
     }
   }
+
+  test("SPARK-10707: nullability should be correctly propagated through set operations") {
+    withTempTable("src") {
+      Seq((1, 1)).toDF("key", "value").registerTempTable("src")
+      checkAnswer(
+        sql("""SELECT count(v) FROM (
+              |  SELECT v FROM (
+              |    SELECT 'foo' AS v FROM src UNION ALL
+              |    SELECT NULL AS v FROM src
+              |  ) my_union WHERE isnull(v)
+              |) my_subview""".stripMargin),
+        Seq(Row(0)))
```

Comment:
Spark SQL does not give you the correct answer now:
```
scala> Seq((1, 1)).toDF("key", "value").registerTempTable("src")

scala> sqlContext.sql("SELECT count(v) FROM (SELECT v FROM (SELECT 'foo' AS v FROM src UNION ALL SELECT NULL AS v FROM src ) my_union WHERE isnull(v) ) my_subview")
res1: org.apache.spark.sql.DataFrame = [_c0: bigint]

scala> res1.collect()
res2: Array[org.apache.spark.sql.Row] = Array([1])
```

And as Mikhail explained in the JIRA, Spark SQL actually gives a different answer when the left and right sides of the UNION are swapped, even though that should be a symmetric operation.